[system] Support callback value in `styleOverrides` slot (#30524)
mui-org/material-ui: docs/src/pages/customization/theme-components/theme-components.md

<p class="description">The theme's components key allows you to customize a component without wrapping it in another component. You can change the styles, the default props, and more.</p>

> ## Default props

Author: moved `Default props` up

To override a lab component's styles with TypeScript, check [this section of the documentation](/components/about-the-lab/#typescript).

### Overrides based on props

You can pass a callback as a value in each slot of the component's `styleOverrides` to apply styles based on props.

The `ownerState` prop is a combination of public props that you pass to the component + internal states inside the component.

```js
const finalTheme = createTheme({
  components: {
    MuiSlider: {
      styleOverrides: {
        valueLabel: ({ ownerState, theme }) => ({
          ...(ownerState.orientation === 'vertical' && {
            backgroundColor: 'transparent',
            color: theme.palette.grey[500],
          }),
        }),
      },
    },
  },
});
```

{{"demo": "pages/customization/theme-components/GlobalThemeOverrideCallback.js"}}

## Adding new component variants

> > ⚠️ This API has been **deprecated** which might be removed in the next major release. If you want to apply styles based on props, take a look at [Overrides based on props](#overrides-based-on-props) instead.

Reviewer: ```suggestion
> ⚠️ This API has been **deprecated** and will likely be removed in the next major release. If you want to apply styles based on props, take a look at [Overrides based on props](#overrides-based-on-props) instead.
```

>

> > To see more details about the why, check out [issue #30412](https://github.com/mui-org/material-ui/issues/30412)

Reviewer: ```suggestion
> If you are interested to see the reasoning behind this change, check out [issue #30412](https://github.com/mui-org/material-ui/issues/30412)
```

To override a lab component's styles with TypeScript, check [this section of the documentation](/components/about-the-lab/#typescript).

### Overrides based on props

You can pass a callback as a value in each slot of the component's `styleOverrides` to apply styles based on props.

> The `ownerState` prop is a combination of public props that you pass to the component + internal states inside the component.

Reviewer 2: ```suggestion
The `ownerState` prop is a combination of public props that you pass to the component + internal state of the component.
```

```js
const finalTheme = createTheme({
  components: {
    MuiSlider: {
      styleOverrides: {
        valueLabel: ({ ownerState, theme }) => ({
          ...(ownerState.orientation === 'vertical' && {
            backgroundColor: 'transparent',
            color: theme.palette.grey[500],
          }),
        }),
      },
    },
  },
});
```

{{"demo": "pages/customization/theme-components/GlobalThemeOverrideCallback.js"}}

### Using `sx` (experimental) syntax

If you are not familiar `sx`, first check out [the concept](/system/the-sx-prop) and [the difference with the `styled`](/system/styled/#difference-with-the-sx-prop).

`sx` is also compatible with theme style overrides if you prefer the shorthand notation.

{{"demo": "pages/customization/theme-components/GlobalThemeOverrideSx.js"}}

## Adding new component variants

> > ⚠️ This API has been **deprecated** and will likely be removed in the next major release. If you want to apply styles based on props, take a look at [Overrides based on props](#overrides-based-on-props) instead.

Reviewer 3: I landed here after seeing the notion of deprecation in #30668. 

We might be jumping one step head too quickly on this one. I'm adding a new comment on why we added the variant API in the past, it doesn't seem that the true purpose in #30412 or in #28107 (from my perspective on why we added it in the first place, it wasn't because of JSS, when we added it we were already adding emotion from what I recall, we knew this PR was going to be possible of it is covered.

Author: Okay, I will remove the deprecation section for now and update the #30412.
